src/verify/mod.rs

```diff
@@ -0,0 +1,309 @@
+            }
+        }
+        self.parent.and_then(|scope| scope.get(name))
+    }
+
+    /// Returns any errors in the given expression when evaluated in this scope.
+    fn check_expr(&'a self, expr: &'b ast::Expr<'b>) -> Vec<Error<'b>> {
+        match &expr.kind {
+            ast::ExprKind::BinOp(left, _, right) => {
+                let mut errors = self.check_expr(&*left);
+                errors.extend(self.check_expr(&*right));
+                errors
+            }
+            ast::ExprKind::PrefixOp(_, inner_expr) => self.check_expr(&*inner_expr),
+            ast::ExprKind::Num(_) => Vec::new(), // Woo! No errors here!
+            // (Overflows will be handled when the int is
+            //  parsed? If the interpreter is going to parse
+            //  from the AST then we can validate here for
+            //  now.)
+            ast::ExprKind::Bracket(block) => self.check_block(&block, 0),
+            ast::ExprKind::Named(name) => match self.get(name.name) {
+                Some(_) => Vec::new(),
+                None => vec![Error {
+                    kind: error::Kind::VariableNotFound,
+                    context: error::Context::Expr,
+                    source: expr.node(),
+                }],
+            },
+            ast::ExprKind::FnCall(fn_expr, args) => {
+                let name = match &fn_expr.kind {
+                    ast::ExprKind::Named(name) => name.name,
+                    _ => {
+                        return vec![Error {
+                            kind: error::Kind::FunctionMustBeName,
+                            context: error::Context::Expr,
+                            source: fn_expr.node(),
+                        }]
+                    }
+                };
+
+                let func = match self.top_level.map.get(name) {
+                    None => {
+                        return vec![Error {
+                            kind: error::Kind::FunctionNotFound,
+                            context: error::Context::Expr,
+                            source: fn_expr.node(),
+                        }]
+                    }
+                    Some(item) => &item.func,
+                };
+
+                if func.n_params != args.len() {
+                    vec![Error {
+                        kind: error::Kind::IncorrectNumberOfArgs {
+                            n_given: args.len(),
+                            n_expected: func.n_params,
+                        },
+                        context: error::Context::Expr,
+                        source: ast::Node::Args(&args),
+                    }]
+                } else {
+                    Vec::new() // Nothing is wrong!!!
+                }
+            }
+            ast::ExprKind::Empty => Vec::new(),
+        }
+    }
+
+    fn check_assign(&'a self, ident: &'b ast::Ident<'b>, expr: &'b ast::Expr<'b>) -> Vec<Error<'b>> {
+        let mut errors = Vec::new();
+        if let None = self.get(ident.name) {
```

Comment:
The idiomatic way of writing this is `if self.get(ident.name).is_none() { ... }` -- [`clippy::style` should produce a warning](https://rust-lang.github.io/rust-clippy/master/#redundant_pattern_matching) for this by default.